File: backend/app/api/routes/buyback.py
"""Selling a used book to the shop, and spending the credit it earns.

The seller's half of the flow. The shop's half — approving, grading what arrives
and paying out — is in `admin.py`, because it needs the admin role and sits with
the rest of the back-office work.

A quote is deliberately free and instant: the whole reason a student sells a course
book here rather than to a classmate is finding out in one click what it is worth.
"""

from __future__ import annotations

from datetime import datetime, timezone
from decimal import Decimal
from uuid import UUID

from datetime import timedelta

from fastapi import APIRouter, Depends, File, Form, HTTPException, Query, UploadFile, status
from sqlalchemy import func
from sqlalchemy.orm import Session, selectinload

from app.core import buyback as pricing
from app.core import storage
from app.core import wallet
from app.core import store_settings
from app.core.config import settings
from app.core.pricing import money
from app.database import db, models
from app.database.dep import require_role
from app.database.models.buyback import (
    BUYBACK_PHOTO_KINDS,
    BuybackPhoto,
    BuybackRequest,
)
from app.schema.pagination import PageMeta
from app.schema.buyback import (
    BuybackDispatch,
    BuybackRequestPage,
    BuybackQuoteRequest,
    BuybackQuoteResponse,
    BuybackRequestRead,
    BuybackSubmitRequest,
    WalletRead,
)

router = APIRouter(prefix="/api/buyback", tags=["Buyback"])

# One seller cannot flood the queue. Not a rate limit on requests per second — a
# cap on how many can be waiting on the shop at once, which is the thing that
# actually costs staff time.
MAX_OPEN_REQUESTS = 10

# Which statuses a seller may still attach or remove photographs on. Once the
# shop has decided, the pictures are evidence of what that decision was taken
# from — changing them afterwards would rewrite the record the decision rests on.
PHOTO_EDITABLE_STATUSES = ("submitted",)


def _quote_expiry():
    """When an offer made now stops standing, or None if expiry is switched off."""
    days = settings.BUYBACK_QUOTE_VALID_DAYS
    if days <= 0:
        return None
    return datetime.now(timezone.utc) + timedelta(days=days)


def _load_own_request(db_session: Session, request_id: UUID, user_id, *, lock: bool = False):
    """One of this seller's requests, or a 404.

    Filtering on `user_id` rather than checking ownership afterwards, so a request
    belonging to somebody else is indistinguishable from one that does not exist —
    a 403 would confirm the id is real.
    """
    query = db_session.query(BuybackRequest).filter(
        BuybackRequest.id == request_id,
        BuybackRequest.user_id == user_id,
    )
    if lock:
        query = query.with_for_update()
    found = query.first()
    if found is None:
        raise HTTPException(status_code=404, detail="Request not found")
    return found


def _price_basis(db_session: Session, book_id, listed_price):
    """Return (title, author, isbn, listed_price) for whichever kind of submission.

    A catalogue book's own price is used rather than anything the client sends: the
    quote is money, so its basis cannot be something the browser gets to choose.
    """
    if book_id is not None:
        book = db_session.query(models.Book).filter(models.Book.id == book_id).first()
        if book is None:
            raise HTTPException(status_code=404, detail="That book is not in our catalogue.")
        if book.condition != "new":
            # A used copy's discounted shelf price is not the printed price, so it
            # is the wrong basis for a quote.
            raise HTTPException(
                status_code=400,
                detail="Pick the main listing for this title, not a used copy.",
            )
        if not book.price or money(book.price) <= 0:
            raise HTTPException(
                status_code=400,
                detail="We do not have a price for that title yet. Enter the price printed on your copy.",
            )
        return book.title, book.author, None, money(book.price)

    return None, None, None, money(listed_price)


@router.post("/quote", response_model=BuybackQuoteResponse)
def get_quote(payload: BuybackQuoteRequest, db: Session = Depends(db.get_db)):
    """What the shop would pay, for every condition at once.

    Public: asking what a book is worth should not require an account. Showing all
    three grades together is what makes a seller grade honestly — they can see what
    admitting to highlighting costs, instead of guessing and risking the offer.
    """
    title, _author, _isbn, listed_price = _price_basis(db, payload.book_id, payload.listed_price)

    try:
        pricing.assert_valid_listed_price(listed_price)
    except pricing.BuybackError as exc:
        raise HTTPException(status_code=400, detail=str(exc))

    options = pricing.quote_breakdown(listed_price, payload.quantity)
    best = max((o["offer"] for o in options), default=Decimal("0.00"))

    return BuybackQuoteResponse(
        title=title,
        listed_price=listed_price,
        quantity=payload.quantity,
        options=options,
        can_sell=best > 0,
        message=(
            None
            if best > 0
            else f"We can only buy books worth {pricing.MIN_BUYBACK_AMOUNT} or more."
        ),
    )


@router.post("", response_model=BuybackRequestRead, status_code=status.HTTP_201_CREATED)
def submit_request(
    payload: BuybackSubmitRequest,
    db: Session = Depends(db.get_db),
    current_user=Depends(require_role("customer")),
):
    """Offer the shop a book.

    The quote is recomputed here rather than accepted from the client, and stored on
    the row. Recomputing stops a browser from naming its own price; storing it means
    a later change to the rates cannot move an offer the seller has already been
    given.
    """
    open_count = (
        db.query(BuybackRequest)
        .filter(
            BuybackRequest.user_id == current_user.id,
            BuybackRequest.status.in_(("submitted", "approved", "received")),
        )
        .count()
    )
    if open_count >= MAX_OPEN_REQUESTS:
        raise HTTPException(
            status_code=409,
            detail=f"You already have {open_count} requests with us. "
                   "Please wait for those to be processed first.",
        )

    title, author, isbn, listed_price = _price_basis(db, payload.book_id, payload.listed_price)

    try:
        offer = pricing.quote(listed_price, payload.condition, payload.quantity)
    except pricing.BuybackError as exc:
        raise HTTPException(status_code=400, detail=str(exc))

    if offer <= 0:
        raise HTTPException(
            status_code=400,
            detail=f"We can only buy books worth {pricing.MIN_BUYBACK_AMOUNT} or more.",
        )

    request = BuybackRequest(
        user_id=current_user.id,
        book_id=payload.book_id,
        title=(title or payload.title or "").strip(),
        author=(author or payload.author or None),
        isbn=(payload.isbn or isbn or None),
        listed_price=listed_price,
        condition=payload.condition,
        quantity=payload.quantity,
        quoted_amount=offer,
        payout_method=payload.payout_method,
        payout_upi=payload.payout_upi,
        payout_account_name=payload.payout_account_name,
        seller_note=payload.seller_note,
        status="submitted",
        quote_expires_at=_quote_expiry(),
    )
    db.add(request)

    # Every admin should see it without polling the queue.
    for admin in db.query(models.User).filter(models.User.role == "admin").all():
        db.add(
            models.Notification(
                user_id=admin.id,
                title="Book offered for buyback",
                body=f'"{request.title}" ({pricing.CONDITION_LABELS[request.condition]}) '
                     f"x{request.quantity} — offer {offer}.",
                is_read=False,
            )
        )

    db.commit()
    db.refresh(request)
    return request


@router.get("", response_model=BuybackRequestPage)
def list_my_requests(
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    db: Session = Depends(db.get_db),
    current_user=Depends(require_role("customer")),
):
    """This seller's requests, newest first.

    Paginated because nothing ever leaves this list: a cancelled or rejected
    request stays forever, and one real account here already carries 148 of them
    — 95 KB on every visit to a page that shows the most recent handful. The
    older ones are still reachable, just not all at once.
    """
    base = db.query(BuybackRequest).filter(BuybackRequest.user_id == current_user.id)
    total = base.with_entities(func.count(BuybackRequest.id)).scalar() or 0
    rows = (
        base
        # Every row renders its photographs, so the default lazy load would be
        # one round trip per request on a page that exists to list them.
        .options(selectinload(BuybackRequest.photos))
        .order_by(BuybackRequest.created_at.desc())
        .offset((page - 1) * page_size)
        .limit(page_size)
        .all()
    )
    open_count = base.filter(
        BuybackRequest.status.in_(("submitted", "approved", "received"))
    ).with_entities(func.count(BuybackRequest.id)).scalar() or 0

    return BuybackRequestPage(
        items=rows,
        meta=PageMeta.from_total(page=page, page_size=page_size, total=total),
        open_count=open_count,
    )


@router.get("/conditions")
def list_conditions():
    """The grades and what each one means.

    Served from the backend so the seller's form and the shop's grading screen
    cannot end up describing the same word differently.
    """
    return {
        "conditions": [
            {
                "value": c,
                "label": pricing.CONDITION_LABELS[c],
                "description": pricing.CONDITION_DESCRIPTIONS[c],
                "buyback_percent": int(pricing.BUYBACK_RATES[c] * 100),
            }
            for c in pricing.CONDITIONS
        ],
        "minimum_amount": pricing.MIN_BUYBACK_AMOUNT,
        # Reported for the same reason as the grades above: the seller's form has
        # to be able to say "we can take two more from you" before the seller
        # picks a book, grades it and chooses a payout — rather than refusing the
        # eleventh request after all that work. A copy of the number in the
        # frontend would be a copy that can drift.
        "max_open_requests": MAX_OPEN_REQUESTS,
        # Where to post an approved book. `None` until configured, which the UI
        # reads as "we will email you the address" rather than telling a seller
        # to post something nowhere.
        "ship_to": settings.buyback_ship_to,
    }


@router.patch("/{request_id}/cancel", response_model=BuybackRequestRead)
def cancel_request(
    request_id: UUID,
    db: Session = Depends(db.get_db),
    current_user=Depends(require_role("customer")),
):
    """Withdraw a request the shop has not paid for yet."""
    request = (
        db.query(BuybackRequest)
        .filter(BuybackRequest.id == request_id, BuybackRequest.user_id == current_user.id)
        .with_for_update()
        .first()
    )
    if request is None:
        raise HTTPException(status_code=404, detail="Request not found")

    if request.status == "cancelled":
        return request
    if request.status in ("paid", "rejected", "expired"):
        raise HTTPException(
            status_code=409,
            detail=f"This request is already {request.status} and cannot be cancelled.",
        )

    request.status = "cancelled"
    request.closed_at = datetime.now(timezone.utc)
    db.commit()
    db.refresh(request)
    return request


# ----- photographs -----

@router.post("/{request_id}/photos", response_model=BuybackRequestRead, status_code=201)
def add_photo(
    request_id: UUID,
    file: UploadFile = File(...),
    kind: str = Form("other"),
    db: Session = Depends(db.get_db),
    current_user=Depends(require_role("customer")),
):
    """Attach a photograph of the book being offered.

    This is what makes the grade mean anything. Without it an admin approved a
    request having never seen the book, re-graded it on arrival, and the payout
    changed — which is the complaint sellers actually make, and it was structural
    rather than anybody's mistake.

    One file per request rather than a multi-file form: a phone upload over a
    patchy connection fails often enough that failing one photograph out of three
    should not lose the other two.
    """
    if kind not in BUYBACK_PHOTO_KINDS:
        raise HTTPException(
            status_code=400,
            detail=f"kind must be one of {', '.join(BUYBACK_PHOTO_KINDS)}",
        )

    request = _load_own_request(db, request_id, current_user.id)
    if request.status not in PHOTO_EDITABLE_STATUSES:
        raise HTTPException(
            status_code=409,
            detail="This request has already been reviewed, so its photos cannot change.",
        )

    if len(request.photos) >= settings.MAX_BUYBACK_PHOTOS:
        raise HTTPException(
            status_code=409,
            detail=f"A request can carry at most {settings.MAX_BUYBACK_PHOTOS} photos.",
        )

    ext = storage.extension_for(file.filename or "")
    if ext is None:
        raise HTTPException(status_code=400, detail="Only .jpg/.jpeg/.png/.webp allowed")

    # Read with a ceiling rather than `file.file.read()`: the unbounded form pulls
    # the whole body into memory, and `Content-Length` is supplied by the client
    # and can lie. Phone photographs are routinely larger than this, which is why
    # the browser downscales before uploading.
    limit = settings.MAX_COVER_UPLOAD_BYTES
    data = file.file.read(limit + 1)
    if not data:
        raise HTTPException(status_code=400, detail="Empty file")
    if len(data) > limit:
        raise HTTPException(
            status_code=413,
            detail=f"Photo must be {limit // (1024 * 1024)}MB or smaller",
        )

    try:
        url = storage.save_buyback_photo(str(request_id), data, ext)
    except storage.StorageError as exc:
        # The upload genuinely did not happen. A 200 here would leave a row
        # pointing at an object that was never written.
        raise HTTPException(status_code=502, detail=f"Could not store the photo: {exc}") from exc

    db.add(
        BuybackPhoto(
            request_id=request.id,
            url=url,
            kind=kind,
            # Append. The seller's order is the order they took them in.
            position=len(request.photos),
        )
    )
    db.commit()
    db.refresh(request)
    return request


@router.delete("/{request_id}/photos/{photo_id}", response_model=BuybackRequestRead)
def remove_photo(
    request_id: UUID,
    photo_id: UUID,
    db: Session = Depends(db.get_db),
    current_user=Depends(require_role("customer")),
):
    """Drop a photograph before the shop has looked at the request.

    The stored file is deliberately left in place. It is content-addressed, so it
    is unreferenced rather than orphaned-and-dangerous, and deleting it here would
    also delete an identical photograph on another request that happens to hash
    the same.
    """
    request = _load_own_request(db, request_id, current_user.id)
    if request.status not in PHOTO_EDITABLE_STATUSES:
        raise HTTPException(
            status_code=409,
            detail="This request has already been reviewed, so its photos cannot change.",
        )

    photo = next((p for p in request.photos if p.id == photo_id), None)
    if photo is None:
        raise HTTPException(status_code=404, detail="Photo not found")

    db.delete(photo)
    db.flush()
    # Close the gap, so `position` stays a dense 0..n-1 and the next upload does
    # not collide with a number a deletion left behind.
    for index, remaining in enumerate(p for p in request.photos if p.id != photo_id):
        remaining.position = index
    db.commit()
    db.refresh(request)
    return request


# ----- dispatch -----

@router.post("/{request_id}/dispatch", response_model=BuybackRequestRead)
def mark_dispatched(
    request_id: UUID,
    payload: BuybackDispatch,
    db: Session = Depends(db.get_db),
    current_user=Depends(require_role("customer")),
):
    """The seller has posted an approved book.

    `approved_at` and `received_at` had nothing between them, so one status —
    "awaiting arrival" — covered both "in the post" and "never sent", and neither
    side could find out which. That is the state a seller chases and the state an
    admin cannot chase, at the same time.

    The consignment number is optional. Plenty of sellers hand a parcel over a
    counter and get no docket, and refusing their "I have posted it" because of
    that would leave the shop blinder than before.
    """
    request = _load_own_request(db, request_id, current_user.id, lock=True)

    if request.status != "approved":
        raise HTTPException(
            status_code=409,
            detail="Only an approved request can be marked as posted.",
        )

    request.seller_tracking_carrier = payload.carrier
    request.seller_tracking_number = payload.tracking_number
    # Idempotent: re-sending to correct a mistyped number must not move the date.
    first_time = request.dispatched_at is None
    if first_time:
        request.dispatched_at = datetime.now(timezone.utc)

    # Only the first time. A seller fixing a digit is not a second parcel, and
    # telling every admin "it is on its way" again turns the correction into
    # noise — the same reason the orders screen's tracking correction is silent.
    if first_time:
        for admin in db.query(models.User).filter(models.User.role == "admin").all():
            db.add(
                models.Notification(
                    user_id=admin.id,
                    title="Book posted for buyback",
                    body=f'"{request.title}" is on its way'
                         + (f" — {request.seller_tracking_number}." if request.seller_tracking_number else "."),
                    is_read=False,
                )
            )

    db.commit()
    db.refresh(request)
    return request


# ----- store credit -----

@router.get("/wallet", response_model=WalletRead)
def get_wallet(
    subtotal: Decimal = Decimal("0"),
    db: Session = Depends(db.get_db),
    current_user=Depends(require_role("customer")),
):
    """Balance, history, and how much of a given subtotal the credit could cover.

    `subtotal` is a query parameter rather than read from the cart so the checkout
    page can ask about the order it is actually about to place, coupon and all.
    """
    return WalletRead(
        balance=wallet.balance(db, current_user.id),
        max_redeemable_now=wallet.max_redeemable(db, current_user.id, subtotal),
        max_redemption_percent=store_settings.current().wallet_max_redemption_percent,
        entries=wallet.history(db, current_user.id),
    )

File: frontend/src/api/buyback.ts
import { getBaseUrl, normalizeCatalogBook, type Book } from "./index";

/** Selling used books to the shop, and the store credit it pays out. */

export type BookCondition = "like_new" | "good" | "fair";

export type ConditionInfo = {
  value: BookCondition;
  label: string;
  description: string;
  buybackPercent: number;
};

export type ConditionOffer = {
  condition: BookCondition;
  label: string;
  description: string;
  offer: number;
  resalePrice: number;
};

export type Quote = {
  title: string | null;
  listedPrice: number;
  quantity: number;
  options: ConditionOffer[];
  /** False when even the best grade falls under the minimum worth handling. */
  canSell: boolean;
  message: string | null;
};

export type BuybackStatus =
  | "submitted"
  | "approved"
  | "received"
  | "paid"
  | "rejected"
  | "cancelled"
  /** The offer's validity window passed before the book was sent. */
  | "expired";

export type SellPhoto = {
  id: string;
  url: string;
  /** cover | spine | damage | other — what it shows, so a set of three reads at
   *  a glance instead of being three pictures of a book. */
  kind: string;
  position: number;
};

export type SellRequest = {
  id: string;
  status: BuybackStatus;
  title: string;
  author: string | null;
  listedPrice: number;
  condition: BookCondition;
  quantity: number;
  quotedAmount: number;
  finalAmount: number | null;
  receivedCondition: BookCondition | null;
  payoutMethod: "wallet" | "bank" | null;
  rejectionReason: string | null;
  /** What the seller photographed. Without these the shop grades a book it has
   *  never seen, re-grades it on arrival, and the payout changes. */
  photos: SellPhoto[];
  /** How the seller posted it, once they did. */
  trackingCarrier: string | null;
  trackingCarrierLabel: string | null;
  trackingNumber: string | null;
  /** Null when the courier has no per-consignment page; the number still shows. */
  trackingUrl: string | null;
  dispatchedAt: string | null;
  /** When the offer stops standing. Null means it does not expire. */
  quoteExpiresAt: string | null;
  createdAt: string;
  paidAt: string | null;
};

export type WalletEntry = {
  id: string;
  amount: number;
  kind: string;
  note: string | null;
  createdAt: string;
};

export type Wallet = {
  balance: number;
  /** How much of the subtotal that was asked about the credit may cover. */
  maxRedeemableNow: number;
  maxRedemptionPercent: number;
  entries: WalletEntry[];
};

/** Surface the API's own message: it is written to be read by the customer. */
const readError = async (res: Response, fallback: string): Promise<string> => {
  const body = await res.json().catch(() => null);
  const detail = (body as { detail?: unknown } | null)?.detail;
  if (typeof detail === "string") return detail;
  if (Array.isArray(detail) && (detail[0] as { msg?: string })?.msg) {
    return String((detail[0] as { msg?: string }).msg);
  }
  return fallback;
};

const authHeaders = (token: string | null) => {
  if (!token) throw new Error("Please sign in to continue.");
  return { "Content-Type": "application/json", Authorization: `Bearer ${token}` };
};

const num = (v: unknown) => (Number.isFinite(Number(v)) ? Number(v) : 0);

// ----- quoting (public) -----

export const fetchConditions = async (): Promise<{
  conditions: ConditionInfo[];
  minimumAmount: number;
  /** How many open requests one seller may have. Reported so it is not copied here. */
  maxOpenRequests: number;
  /** Where to post an approved book, as address lines. Null until configured. */
  shipTo: string[] | null;
}> => {
  const res = await fetch(`${getBaseUrl()}/api/buyback/conditions`);
  if (!res.ok) throw new Error("Could not load the condition guide");
  const raw = await res.json();
  return {
    conditions: (raw?.conditions ?? []).map((c: any) => ({
      value: c.value,
      label: c.label,
      description: c.description,
      buybackPercent: num(c.buyback_percent),
    })),
    minimumAmount: num(raw?.minimum_amount),
    maxOpenRequests: num(raw?.max_open_requests) || 0,
    shipTo: Array.isArray(raw?.ship_to) && raw.ship_to.length ? raw.ship_to : null,
  };
};

export const fetchQuote = async (params: {
  bookId?: string;
  listedPrice?: number | string;
  quantity?: number;
}): Promise<Quote> => {
  const res = await fetch(`${getBaseUrl()}/api/buyback/quote`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      book_id: params.bookId ?? null,
      listed_price: params.listedPrice ?? null,
      quantity: params.quantity ?? 1,
    }),
  });
  if (!res.ok) throw new Error(await readError(res, "Could not price that book"));

  const raw = await res.json();
  return {
    title: raw?.title ?? null,
    listedPrice: num(raw?.listed_price),
    quantity: num(raw?.quantity) || 1,
    options: (raw?.options ?? []).map((o: any) => ({
      condition: o.condition,
      label: o.label,
      description: o.description,
      offer: num(o.offer),
      resalePrice: num(o.resale_price),
    })),
    canSell: Boolean(raw?.can_sell),
    message: raw?.message ?? null,
  };
};

// ----- submitting and tracking -----

const normalizeRequest = (raw: any): SellRequest => ({
  id: String(raw?.id ?? ""),
  status: raw?.status ?? "submitted",
  title: String(raw?.title ?? ""),
  author: raw?.author ?? null,
  listedPrice: num(raw?.listed_price),
  condition: raw?.condition ?? "good",
  quantity: num(raw?.quantity) || 1,
  quotedAmount: num(raw?.quoted_amount),
  finalAmount: raw?.final_amount == null ? null : num(raw.final_amount),
  receivedCondition: raw?.received_condition ?? null,
  payoutMethod: raw?.payout_method ?? null,
  rejectionReason: raw?.rejection_reason ?? null,
  photos: Array.isArray(raw?.photos)
    ? raw.photos.map((p: any) => ({
        id: String(p?.id ?? ""),
        url: String(p?.url ?? ""),
        kind: String(p?.kind ?? "other"),
        position: num(p?.position),
      }))
    : [],
  trackingCarrier: raw?.seller_tracking_carrier ?? null,
  trackingCarrierLabel: raw?.seller_tracking_carrier_label ?? null,
  trackingNumber: raw?.seller_tracking_number ?? null,
  trackingUrl: raw?.seller_tracking_url ?? null,
  dispatchedAt: raw?.dispatched_at ?? null,
  quoteExpiresAt: raw?.quote_expires_at ?? null,
  createdAt: String(raw?.created_at ?? ""),
  paidAt: raw?.paid_at ?? null,
});

type SubmitSellRequest = {
  bookId?: string;
  title?: string;
  author?: string;
  isbn?: string;
  listedPrice?: number | string;
  condition: BookCondition;
  quantity: number;
  payoutMethod: "wallet" | "bank";
  payoutUpi?: string;
  payoutAccountName?: string;
  sellerNote?: string;
};

export const submitSellRequest = async (
  token: string | null,
  input: SubmitSellRequest,
): Promise<SellRequest> => {
  const res = await fetch(`${getBaseUrl()}/api/buyback`, {
    method: "POST",
    headers: authHeaders(token),
    body: JSON.stringify({
      book_id: input.bookId ?? null,
      title: input.title ?? null,
      author: input.author ?? null,
      isbn: input.isbn ?? null,
      listed_price: input.listedPrice ?? null,
      condition: input.condition,
      quantity: input.quantity,
      payout_method: input.payoutMethod,
      payout_upi: input.payoutUpi ?? null,
      payout_account_name: input.payoutAccountName ?? null,
      seller_note: input.sellerNote ?? null,
    }),
  });
  if (!res.ok) throw new Error(await readError(res, "Could not submit your book"));
  return normalizeRequest(await res.json());
};

type SellRequestPage = {
  items: SellRequest[];
  meta: { page: number; page_size: number; total: number; pages: number };
  /** How many are still with the shop, across the whole list — the sell form
   *  warns against the per-seller cap with this. Not a count of `items`. */
  openCount: number;
};

/** This seller's requests, newest first.
 *
 * Paginated because nothing ever leaves the list — a cancelled or rejected
 * request stays — and one real account already carries 148 of them, which was
 * 95 KB on every visit.
 */
export const fetchMySellRequests = async (
  token: string | null,
  params?: { page?: number; pageSize?: number },
): Promise<SellRequestPage> => {
  const query = new URLSearchParams();
  if (params?.page) query.set("page", String(params.page));
  if (params?.pageSize) query.set("page_size", String(params.pageSize));
  const res = await fetch(
    `${getBaseUrl()}/api/buyback${query.toString() ? `?${query.toString()}` : ""}`,
    { headers: authHeaders(token) },
  );
  if (!res.ok) throw new Error(await readError(res, "Could not load your sell requests"));
  const raw = await res.json();
  return {
    items: (raw?.items ?? []).map(normalizeRequest),
    meta: raw?.meta ?? { page: 1, page_size: 20, total: 0, pages: 0 },
    openCount: num(raw?.open_count),
  };
};

export const cancelSellRequest = async (
  token: string | null,
  requestId: string,
): Promise<SellRequest> => {
  const res = await fetch(`${getBaseUrl()}/api/buyback/${requestId}/cancel`, {
    method: "PATCH",
    headers: authHeaders(token),
  });
  if (!res.ok) throw new Error(await readError(res, "Could not cancel that request"));
  return normalizeRequest(await res.json());
};

// ----- store credit -----

export const fetchWallet = async (
  token: string | null,
  subtotal = 0,
): Promise<Wallet> => {
  const res = await fetch(
    `${getBaseUrl()}/api/buyback/wallet?subtotal=${encodeURIComponent(subtotal)}`,
    { headers: authHeaders(token) },
  );
  if (!res.ok) throw new Error(await readError(res, "Could not load your credit"));

  const raw = await res.json();
  return {
    balance: num(raw?.balance),
    maxRedeemableNow: num(raw?.max_redeemable_now),
    maxRedemptionPercent: num(raw?.max_redemption_percent),
    entries: (raw?.entries ?? []).map((e: any) => ({
      id: String(e.id),
      amount: num(e.amount),
      kind: String(e.kind ?? ""),
      note: e.note ?? null,
      createdAt: String(e.created_at ?? ""),
    })),
  };
};

// ----- used copies of a title -----

export const fetchUsedCopies = async (bookId: string): Promise<Book[]> => {
  const res = await fetch(`${getBaseUrl()}/api/catalog/books/${bookId}/used`);
  // A missing used-copies strip must never break the product page around it.
  if (!res.ok) return [];

  const raw = await res.json();
  return (Array.isArray(raw) ? raw : []).map((b) => normalizeCatalogBook(b));
};

/** How a wallet ledger `kind` reads to the customer. */
export const WALLET_KIND_LABELS: Record<string, string> = {
  buyback_payout: "Sold a book",
  order_redemption: "Used on an order",
  order_refund: "Returned from a cancelled order",
  adjustment: "Adjustment",
};

// ---------------------------------------------------------------------------
// Photographs
// ---------------------------------------------------------------------------

/**
 * Shrink a photograph before it leaves the phone.
 *
 * A modern phone camera produces 4–8 MB per shot and the server's ceiling is
 * 5 MB, so without this a seller's first upload fails and they have no idea why.
 * 1600px on the long edge is more than enough to judge a crease or a cracked
 * spine, and it turns a 6 MB upload over patchy mobile data into ~300 KB.
 *
 * Falls back to the original file if anything in the canvas path fails: a photo
 * that uploads slowly is better than one that does not upload.
 */
export const downscaleImage = async (file: File, maxEdge = 1600): Promise<File> => {
  try {
    const bitmap = await createImageBitmap(file);
    const scale = Math.min(1, maxEdge / Math.max(bitmap.width, bitmap.height));
    if (scale === 1 && file.size <= 1_500_000) return file;

    const canvas = document.createElement("canvas");
    canvas.width = Math.round(bitmap.width * scale);
    canvas.height = Math.round(bitmap.height * scale);
    const ctx = canvas.getContext("2d");
    if (!ctx) return file;
    ctx.drawImage(bitmap, 0, 0, canvas.width, canvas.height);

    const blob = await new Promise<Blob | null>((resolve) =>
      canvas.toBlob(resolve, "image/jpeg", 0.82),
    );
    if (!blob || blob.size >= file.size) return file;
    // The server decides what it accepts from the *extension*, so the name has
    // to match the type the canvas just produced.
    return new File([blob], file.name.replace(/\.[^.]+$/, "") + ".jpg", { type: "image/jpeg" });
  } catch {
    return file;
  }
};

export const uploadSellPhoto = async (
  token: string | null,
  requestId: string,
  file: File,
  kind: "cover" | "spine" | "damage" | "other",
): Promise<SellRequest> => {
  const form = new FormData();
  form.append("file", await downscaleImage(file));
  form.append("kind", kind);
  const res = await fetch(`${getBaseUrl()}/api/buyback/${requestId}/photos`, {
    method: "POST",
    headers: authHeaders(token),
    body: form,
  });
  if (!res.ok) throw new Error(await readError(res, "Could not upload that photo"));
  return normalizeRequest(await res.json());
};

export const deleteSellPhoto = async (
  token: string | null,
  requestId: string,
  photoId: string,
): Promise<SellRequest> => {
  const res = await fetch(`${getBaseUrl()}/api/buyback/${requestId}/photos/${photoId}`, {
    method: "DELETE",
    headers: authHeaders(token),
  });
  if (!res.ok) throw new Error(await readError(res, "Could not remove that photo"));
  return normalizeRequest(await res.json());
};

/** Tell the shop the book is in the post.
 *
 * The consignment number is optional: plenty of sellers hand a parcel over a
 * counter and get no docket, and refusing their "I have posted it" because of
 * that would leave the shop blinder than before.
 */
export const markSellDispatched = async (
  token: string | null,
  requestId: string,
  tracking?: { carrier: string; trackingNumber: string },
): Promise<SellRequest> => {
  const res = await fetch(`${getBaseUrl()}/api/buyback/${requestId}/dispatch`, {
    method: "POST",
    headers: { ...authHeaders(token), "Content-Type": "application/json" },
    body: JSON.stringify(
      tracking ? { carrier: tracking.carrier, tracking_number: tracking.trackingNumber } : {},
    ),
  });
  if (!res.ok) throw new Error(await readError(res, "Could not record the dispatch"));
  return normalizeRequest(await res.json());
};
